lib/OpenQA/Scheduler/Scheduler.pm

```diff
@@ -138,17 +150,397 @@ sub _prefer_parallel {
+                    OpenQA::Scheduler::BUSY_BACKOFF() ?
+                      OpenQA::Scheduler::SCHEDULE_TICK_MS() + 1000
+                    : OpenQA::Scheduler::SCHEDULE_TICK_MS());
+            })) if (OpenQA::Scheduler::CONGESTION_CONTROL());
+
+    my @allocated_jobs;
+
+    log_debug("-> Scheduling new jobs.");
+    try {
+        @allocated_jobs = schema->txn_do(
+            sub {
+                my $all_workers = schema->resultset("Workers")->count();
+                my @allocated_workers;
+                my @allocated_jobs;
+
+                # $worker->connected is too much expensive, prefeer dead.
+                my @free_workers = grep { !$_->dead } schema->resultset("Workers")->search({job_id => undef},)->all();
+                @free_workers = shuffle(@free_workers)
+                  if OpenQA::Scheduler::SHUFFLE_WORKERS();   # shuffle avoids starvation if a free worker keeps failing.
+                log_debug("\t Free workers: " . scalar(@free_workers) . "/$all_workers");
+                log_debug("\t Failure# ${failure}") if OpenQA::Scheduler::CONGESTION_CONTROL();
+
+                # Consider it a failure if OpenQA::Scheduler::CONGESTION_CONTROL is set
+                # so if there are no free workers scheduler will kick in later.
+                $failure++
+                  and return ()
+                  if @free_workers == 0
+                  && OpenQA::Scheduler::CONGESTION_CONTROL()
+                  && OpenQA::Scheduler::BUSY_BACKOFF()
+                  && $all_workers > 0;
+                my $allocating = {};
+                for my $w (@free_workers) {
+
+                    my @possible_jobs = job_grab(
+                        workerid     => $w->id(),
+                        blocking     => 0,
+                        allocate     => 0,
+                        scheduler    => 1,
+                        jobs         => OpenQA::Scheduler::MAX_JOB_ALLOCATION(),
+                        max_attempts => OpenQA::Scheduler::FIND_JOB_ATTEMPTS());
+                    log_debug("[Worker#" . $w->id() . "] Possible jobs to allocate: " . scalar(@possible_jobs));
+                    my $allocated_job;
+                    for my $p_job (@possible_jobs) {
+                        next if exists $p_job->{id} && exists $allocating->{$p_job->{id}};
+                        last if $allocated_job;
+                        $allocated_job = $p_job;
+                    }
+                    log_debug(pp(\@possible_jobs));
+                    log_debug("[Worker#" . $w->id() . "] Among them, we have chosen job: " . $allocated_job->{id})
+                      if $allocated_job && exists $allocated_job->{id};
+
+                    $allocated_job->{assigned_worker_id} = $w->id() if $allocated_job;
+                    next unless $allocated_job && exists $allocated_job->{id};
+
+
+      # TODO: we need to be sure job_grab is not returning the same job for different workers - for now do not push them
```

Comment:
is this a complain wrt perltidy? 
you could blame me for putting that in one single line 😄